Modules/posixmodule.c

```diff
@@ -11420,14 +11421,50 @@ os_lseek_impl(PyObject *module, int fd, Py_off_t position, int how)
     Py_BEGIN_ALLOW_THREADS
     _Py_BEGIN_SUPPRESS_IPH
 #ifdef MS_WINDOWS
-    result = _lseeki64(fd, position, how);
+    switch (how) {
+    case SEEK_SET:
+        how = FILE_BEGIN;
+        break;
+    case SEEK_CUR:
+        how = FILE_CURRENT;
+        break;
+    case SEEK_END:
+        how = FILE_END;
+        break;
+    }
+    HANDLE h = (HANDLE)_get_osfhandle(fd);
+    if (h == INVALID_HANDLE_VALUE) {
+        result = -1;
+    }
+    if (result >= 0) {
+        if (GetFileType(h) != FILE_TYPE_DISK) {
+            // Only file is seekable
+            errno = ESPIPE;
+            result = -1;
+        }
+    }
+    if (result >= 0) {
+        LARGE_INTEGER distance, newdistance;
+        distance.QuadPart = position;
+        if (SetFilePointerEx(h, distance, &newdistance, how)) {
+            result = newdistance.QuadPart;
+        } else {
+            result = -1;
+        }
+    }
 #else
     result = lseek(fd, position, how);
 #endif
     _Py_END_SUPPRESS_IPH
     Py_END_ALLOW_THREADS
-    if (result < 0)
+    if (result < 0) {
+#ifdef MS_WINDOWS
+        if (errno == 0) {
+            errno = winerror_to_errno(GetLastError());
+        }
```

Comment:
Why not use `PyErr_SetFromWindowsErr(0)`?

Reply:
`SetFilePointerEx` will set the error to `GetLastError` instead of `errno`, so we should check it. But after changed to call `_lseeki64` directly, there is no need for this line of code.